tests/gpflow/utilities/test_ops.py

```diff
@@ -0,0 +1,45 @@
+import pytest
+import numpy as np
+import tensorflow as tf
+
+import gpflow
+
+
+def pca_reduce(X: np.ndarray, Q: np.ndarray) -> np.ndarray:
+    """
+    A helpful function for linearly reducing the dimensionality of the data X
+    to Q.
+    :param X: data array of size N (number of points) x D (dimensions)
+    :param Q: Number of latent dimensions, Q < D
+    :return: PCA projection array of size N x Q.
+    """
+    if Q > X.shape[1]:  # pragma: no cover
+        raise ValueError("Cannot have more latent dimensions than observed")
+    if isinstance(X, tf.Tensor):
+        X = X.numpy()
+        # TODO why not use tf.linalg.eigh?
+    evals, evecs = np.linalg.eigh(np.cov(X.T))
+    W = evecs[:, -Q:]
+    return (X - X.mean(0)).dot(W)
+
+
+@pytest.mark.parametrize("N", [3, 7])
+@pytest.mark.parametrize("D", [2, 5, 9])
+@pytest.mark.parametrize("Q", [2, 5, 9])
+def test_numpy_equivalence(N, D, Q):
+    X = np.random.randn(N, D)
+
+    if Q > D:
```

Comment:
nit: this looks a bit ugly - how about splitting the exception case (Q > D) to another unittest.

Reply:
true, but makes it harder to parametrize (you can't parametrize conditionally, so then I'd have to write out the whole set of combinations instead of two single lists)